### PR TITLE
Setting up ASV to add benchmarks

### PR DESCRIPTION
**Pull request by OmarWagih1**
In the discussion in this issue #522 the package is aiming to become more performance efficient and so we should start benchmarking our functions to have a baseline of where we are at, this issue starts an effort in benchmarking by setting up a benchmarks folder that uses ASV.

Currently in the benchmarking folder there is a Readme that allows people to use ASV and guides them to where to go to understand how to write benchmarks along with a configuration file to run the benchmarks both on Pip and Conda.

This is accompanied by a simple benchmark taken from the online tutorial for the sake of display, more benchmarks to come next and in the GSoC project.

### CODE REVIEW

File: benchmarks/asv.conf.venv.json [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    // "html_dir": "html",
+
+    // The number of characters to retain in the commit hashes.
+    // "hash_length": 8,
+
+    // `asv` will cache wheels of the recent builds in each
+    // environment, making them faster to install next time.  This is
+    // number of builds to keep, per environment.
+    // "wheel_cache_size": 0
+}
```
Comment (KelSolaar) [inline benchmarks/asv.conf.venv.json]: Codacy Issue found: [Unexpected character ('}' (code 125)): was expecting double-quote to start field name](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5223583)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- benchmarks/benchmarks/benchmarks.py  1
         

Clones added
============
- benchmarks/benchmarks/benchmarks.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5223583&bid=17125567)

File: benchmarks/asv.conf.conda.json [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    // "html_dir": "html",
+
+    // The number of characters to retain in the commit hashes.
+    // "hash_length": 8,
+
+    // `asv` will cache wheels of the recent builds in each
+    // environment, making them faster to install next time.  This is
+    // number of builds to keep, per environment.
+    // "wheel_cache_size": 0
+}
```
Comment (KelSolaar) [inline benchmarks/asv.conf.conda.json]: Codacy Issue found: [Unexpected character ('}' (code 125)): was expecting double-quote to start field name](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5223583)

File: benchmarks/asv.conf.json [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+    // The number of characters to retain in the commit hashes.
+    // "hash_length": 8,
+
+    // `asv` will cache results of the recent builds in each
+    // environment, making them faster to install next time.  This is
+    // the number of builds to keep, per environment.
+    // "build_cache_size": 2,
+
+}
```
Comment (KelSolaar) [inline benchmarks/asv.conf.json]: Codacy Issue found: [Unexpected character ('}' (code 125)): was expecting double-quote to start field name](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5223583)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- benchmarks/benchmarks/benchmarks.py  1
         

Clones added
============
- benchmarks/benchmarks/benchmarks.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5223583&bid=17125567)

**Comment by KelSolaar** [comment]
Thanks @OmarWagih1!

Exciting to see the first step on that! Paging @MichaelMauderer so that he can drop an eye!

I was looking at https://github.com/sympy/sympy_benchmarks and https://github.com/astropy/astropy-benchmarks which made me understand your suggestion for having a separate repo, I haven't used ASV but I was assuming it could discover benchmarks like Unittest/Nose/Pytest but this is not the case, they need to live in a separate tree. No big deal!

So while I still think that there is value having them living in the Colour repo next to the code so that we keep an eye on them that being said something I did not consider is that if we run them during CI, it will increase the already long CI time so... ...it might not be that of a good idea to have them in the same repo (or we would need to run those on a separate process, maybe successful CI could trigger another GH action that would run the benchmarks specifically).

I also see that the directory structure you drafted is `benchmarks/benchmarks/.*py` because well, you need to namespace it in the **Colour** repo and then you adopt the **ASV** default directory which is `benchmarks`.

Given the **Sympy** and **Astropy** precedent, my thinking has back-pedalled and maybe a separate repository, e.g. `colour-benchmarks` is a good idea!

Thoughts?

**Comment by MichaelMauderer** [comment]
My only concern with the external repo would be that it might be harder to keep it up to date, as changes in the main repo also need to update the benchmarking repo.
Using a separate CI action that is independent of the tests and runs with different triggers might be easier to start with.

I guess the real question might be, when do we want to run the benchmarks? For every PR? For releases? Or only for specific performance features?

**Comment by MichaelMauderer** [comment]
But we can just start it off in a separate repo and later on decide on whether we want to integrate it.

**Comment by OmarWagih1** [comment]
@MichaelMauderer Is it going to be harder because you'd need to do two pull requests on two different repos to update your code and the benchmark for it? If so then yes, I agree.

We could run benchmarks on every commit as long as the benchmark suite is not huge and the resources it take are not massive, for example my test benchmark took about 16 micro seconds, so even if the average test is ten times that, it wouldn't take much to benchmark on every commit.

**Comment by KelSolaar** [comment]
Closing this one in favour of the dedicated repository here: https://github.com/colour-science/colour-benchmarks

**Pull request closed without merging**